lib/fluent/plugin/out_forward/socket_cache.rb

```diff
@@ -19,135 +19,119 @@
 module Fluent::Plugin
   class ForwardOutput < Output
     class SocketCache
-      TimedSocket = Struct.new(:timeout, :sock, :ref)
+      TimedSocket = Struct.new(:timeout, :key, :sock)
 
       def initialize(timeout, log)
         @log = log
         @timeout = timeout
-        @active_socks = {}
-        @inactive_socks = {}
+        @available_sockets = Hash.new { |obj, k| obj[k] = [] }
+        @inflight_sockets = {}
+        @inactive_sockets = []
         @mutex = Mutex.new
       end
 
-      def revoke(key = Thread.current.object_id)
+      def checkout_or(key)
         @mutex.synchronize do
-          if @active_socks[key]
-            @inactive_socks[key] = @active_socks.delete(key)
-            @inactive_socks[key].ref = 0
+          tsock = pick_socket(key)
+
+          if tsock
+            tsock.sock
+          else
+            val = yield
```

Comment:
`sock` or `new_sock` is better than `val`

Reply:
fixed https://github.com/fluent/fluentd/pull/2501/commits/14985b352e909c984f26e44253bcdff13116a0ff